homeassistant/components/modbus/binary_sensor.py

```diff
@@ -79,15 +81,23 @@ def device_class(self) -> Optional[str]:
         """Return the device class of the sensor."""
         return self._device_class
 
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._available
+
     def update(self):
         """Update the state of the sensor."""
-        result = self._hub.read_coils(self._slave, self._coil, 1)
         try:
+            result = self._hub.read_coils(self._slave, self._coil, 1)
             self._value = result.bits[0]
-        except AttributeError:
-            _LOGGER.error(
-                "No response from hub %s, slave %s, coil %s",
-                self._hub.name,
-                self._slave,
-                self._coil,
-            )
+            self._available = True
+        except ConnectionException:
```

Comment:
You're no longer catching `AttributeError`. What used to raise it? If it was because `result` was `None`, check for that in an if-statement.

Reply:
I had to look at the `pymodbus` library more thoroughly. From my point of view, the original check for the `AttributeError` is wrong, and insufficient. You can get this error only if the Modbus connection is established, and the Modbus server sends an invalid response, in which case the log entry claiming `no response from the hub` is incorrect.

PyModbus works like this:
- It raises 5 kinds of exceptions; we could in theory receive `ModbusIOException`, `ParameterException`, `NoSuchSlaveException`, `ConnectionException`, and `InvalidMessageReceivedException`
- Exceptions `ParameterException` and `NoSuchSlaveException` can happen due to misconfiguration in our config, or because the Modbus server doesn't respond (we request data from slave 2 and no device responds)
- In all other cases, we receive a response. If the Modbus server is misconfigured, e.g., its registers don't hold data at the requested address, `pymodbus` returns an `ExceptionResponse`.

I suggest to:
- Check for `ModbusException`, which will cover all the exceptions mentioned above. They all extend this class.
- Check explicitly for an `ExceptionResponse`, instead of trying to get attributes from response and getting `AttributeError`

I'm open to any suggestions.

Thanks!

Reply:
Your suggestion sounds good! 👍

Reply:
I implemented latest changes. A couple of notes:
- `pymodbus` raises only the `ConnectionException`, all other exceptions are returned as a normal response. I introduced a check which tells if result object is either a `ModbusException` or an `ExceptionResponse`.
- Such mechanism was introduced for all existing Modbus entities.
- Each entity was tested manually - I tried to simulate loss of connectivity, and accessing Modbus address outside of defined scope. In both cases, I log the error as "No response from hub", which may not be the exact case, but it should be clear enough that there is an issue which needs to be checked manually (either misconfiguration or connectivity issue)
- UI correctly displays an entity as unavailable if there's an issue
- This change elliminates most of the abusive prints into the log file. However, `pymodbus` itself logs connectivity issues each time it's called. There's probably nothing we can do about it at the moment. Maybe we could fix this behavior by reworking Home Assistant code to use their Async API.

Here's an output from log for illustration. Note that `pymodbus` logs an error each two seconds. At least home assistant logs this error only once. It's an improvement already:
```bash
2020-02-11 14:38:15 WARNING (MainThread) [homeassistant.components.switch] Updating modbus switch took longer than the scheduled update interval 0:00:01
2020-02-11 14:38:15 ERROR (SyncWorker_5) [homeassistant.components.modbus.switch] No response from hub plc_01, slave 1, register 0
2020-02-11 14:38:17 ERROR (SyncWorker_4) [pymodbus.client.sync] Connection to (127.0.0.1, 5020) failed: [Errno 111] Connection refused
2020-02-11 14:38:19 ERROR (SyncWorker_2) [pymodbus.client.sync] Connection to (127.0.0.1, 5020) failed: [Errno 111] Connection refused
2020-02-11 14:38:21 ERROR (SyncWorker_6) [pymodbus.client.sync] Connection to (127.0.0.1, 5020) failed: [Errno 111] Connection refused
2020-02-11 14:38:23 ERROR (SyncWorker_3) [pymodbus.client.sync] Connection to (127.0.0.1, 5020) failed: [Errno 111] Connection refused
```